backend/src/main/java/fk/prof/backend/model/assignment/AggregationWindowPlanner.java

```diff
@@ -0,0 +1,203 @@
+              .setDescription(latestWorkProfile.getDescription())
+              .setDuration(latestWorkProfile.getDuration());
+
+          workAssignmentBuilders[i] = workAssignmentBuilder;
+          workIds[i] = workAssignmentBuilder.getWorkId();
+        }
+        setupAggregationWindow(workAssignmentBuilders, workIds);
+      } catch (Exception ex) {
+        reset();
+        //TODO: log this as metric somewhere, fatal failure wrt to aggregation window
+        logger.error("Skipping work assignments and setup of aggregation window because of unexpected error while processing for process_group=" + RecorderProtoUtil.processGroupCompactRepr(processGroup));
+      }
+    } else {
+      //TODO: log this as metric somewhere, fatal failure wrt to aggregation window
+      logger.error("Skipping work assignments and setup of aggregation window because work profile was not fetched in time for process_group=" + RecorderProtoUtil.processGroupCompactRepr(processGroup));
+    }
+
+    vertx.setTimer(((aggregationWindowDurationInMins * 60) - workProfileRefreshBufferInSecs) * MILLIS_IN_SEC,
+        timerId -> getWorkForNextAggregationWindow(currentAggregationWindowIndex + 1));
+  }
+
+  private void setupAggregationWindow(Recorder.WorkAssignment.Builder[] workAssignmentBuilders, long[] workIds) {
+    LocalDateTime windowStart = LocalDateTime.now(Clock.systemUTC());
+    int maxConcurrentSlotsForScheduling = simultaneousWorkAssignmentCounter.acquireSlots(processGroup, latestWorkProfile);
+    try {
+      WorkAssignmentSchedule workAssignmentSchedule = workAssignmentScheduleFactory.getNewWorkAssignmentSchedule(workAssignmentBuilders,
+          latestWorkProfile.getCoveragePct(),
+          maxConcurrentSlotsForScheduling,
+          latestWorkProfile.getDuration());
+      currentlyOccupiedWorkAssignmentSlots = workAssignmentSchedule.getMaxConcurrentlyScheduledEntries();
+      currentAggregationWindow = new AggregationWindow(
+          processGroup.getAppId(),
+          processGroup.getCluster(),
+          processGroup.getProcName(),
+          windowStart,
+          workIds);
+      processGroupContextForScheduling.updateWorkAssignmentSchedule(workAssignmentSchedule);
+    } catch (Exception ex) {
+      //Any exception thrown is first detected here so that acquired slots can be released and then propagated further up the chain so that caller for setting up aggregation window can react to it
+      simultaneousWorkAssignmentCounter.releaseSlots(maxConcurrentSlotsForScheduling);
+      throw ex;
+    }
+    simultaneousWorkAssignmentCounter.releaseSlots(maxConcurrentSlotsForScheduling - currentlyOccupiedWorkAssignmentSlots);
+    aggregationWindowLookupStore.associateAggregationWindow(workIds, currentAggregationWindow);
+  }
+
+  private void expireCurrentAggregationWindow() {
+    if(currentAggregationWindow != null) {
+      simultaneousWorkAssignmentCounter.releaseSlots(currentlyOccupiedWorkAssignmentSlots);
```

Comment:
Fixed in commit https://github.com/Flipkart/fk-prof/commit/14c6b7773d38b356a1a8daa5cf5a98e8ef5862ab